File: src/main/java/com/chattriggers/ctjs/commands/Command.java
package com.chattriggers.ctjs.commands;

import com.chattriggers.ctjs.minecraft.libs.ChatLib;
import com.chattriggers.ctjs.triggers.OnTrigger;
import com.chattriggers.ctjs.utils.console.Console;
import lombok.Getter;
import lombok.Setter;
import net.minecraft.command.CommandBase;
import net.minecraft.command.CommandException;
import net.minecraft.command.ICommandSender;
import net.minecraft.server.MinecraftServer;
import net.minecraft.util.BlockPos;

import java.util.ArrayList;
import java.util.List;

public class Command extends CommandBase {
    private String name;
    private String usage;
    @Setter
    private List<String> tabComplete;
    @Getter
    private ArrayList<OnTrigger> triggers = new ArrayList<>();

    public Command(OnTrigger trigger, String name, String usage) {
        this.triggers.add(trigger);
        this.name = name;
        this.usage = usage;
    }

    public void addTabComplete(String option) {
        this.tabComplete.add(option);
    }

    public String getName() {
        return getCommandName();
    }
    public String getCommandName() {
        return name;
    }

    public List<String> addTabCompletionOptions(ICommandSender sender, String[] args, BlockPos pos) {
        return tabComplete;
    }

    public int getRequiredPermissionLevel() {
        return 0;
    }

    public String getUsage(ICommandSender sender) {
        return getCommandUsage(sender);
    }
    public String getCommandUsage(ICommandSender sender) {
        return usage;
    }

    public void execute(MinecraftServer server,ICommandSender sender, String[] args) throws CommandException {
        processCommand(sender, args);
    }
    public void processCommand(ICommandSender sender, String[] args) throws CommandException {
        try {
            for (OnTrigger trigger : triggers)
                trigger.trigger((Object[]) args);
        } catch (Exception exception) {
            ChatLib.chat("&cSomething went wrong while running that command");
            ChatLib.chat("&cCheck the ct console for more information");
            Console.getInstance().printStackTrace(exception);
        }
    }
}

File: src/main/kotlin/com/chattriggers/ctjs/commands/Command.kt
package com.chattriggers.ctjs.commands

import com.chattriggers.ctjs.printToConsole
import com.chattriggers.ctjs.triggers.Trigger
import com.chattriggers.ctjs.utils.console.LogType
import net.minecraft.command.CommandBase
import net.minecraft.command.CommandException
import net.minecraft.command.ICommandSender
import net.minecraft.util.BlockPos

//#if MC==10809
import net.minecraftforge.client.ClientCommandHandler
//#endif

class Command(
    private val trigger: Trigger,
    private val name: String,
    private val usage: String,
    private val tabCompletionOptions: MutableList<String>,
    private var aliases: MutableList<String>,
    private val overrideExisting: Boolean = false,
    private val callback: ((Array<out String>) -> MutableList<String>)? = null,
) : CommandBase() {

    fun getTriggers() = listOf(trigger)

    //#if MC<=10809
    override fun getCommandName() = name
    //#else
    //$$ override fun getName() = name
    //#endif

    //#if MC<=10809
    override fun getCommandAliases() = aliases
    //#else
    //$$ override fun getAliases() = aliases
    //#endif

    override fun getRequiredPermissionLevel() = 0

    //#if MC<=10809
    override fun getCommandUsage(sender: ICommandSender) = usage
    //#else
    //$$ override fun getUsage(sender: ICommandSender) = usage
    //#endif

    override fun addTabCompletionOptions(
        sender: ICommandSender?,
        args: Array<out String>?,
        pos: BlockPos?,
    ): MutableList<String> {
        return callback?.invoke(args ?: arrayOf())?.toMutableList() ?: tabCompletionOptions
    }

    @Throws(CommandException::class)
    //#if MC<=10809
    override fun processCommand(sender: ICommandSender, args: Array<String>) = trigger.trigger(args)
    //#else
    //$$ override fun execute(server: net.minecraft.server.MinecraftServer?, sender: ICommandSender, args: Array<String>) = trigger(args)
    //#endif

    fun register() {
        if (name in ClientCommandHandler.instance.commands.keys && !overrideExisting) {
            ("Command with name $name already exists! " +
                    "This will not override the other command with the same name. " +
                    "To override conflicting commands, " +
                    "set the 2nd argument in setName() to true.").printToConsole(trigger.loader.console, LogType.WARN)

            return
        }

        ClientCommandHandler.instance.registerCommand(this)
        activeCommands[name] = this
    }

    fun unregister() {
        ClientCommandHandler.instance.commandSet.remove(this)
        ClientCommandHandler.instance.commands.remove(name)
        activeCommands.remove(name)
    }

    companion object {
        internal val activeCommands = mutableMapOf<String, Command>()
    }
}
